### PR TITLE
RedeemPool, Redeemer proxy contract and tests

### LINKED ISSUE #788: Redeem composer contract
This is a replica of OSTComposer, expect that this is for redeeming.

### PR DESCRIPTION
**Pull request by sarvesh-ost**
fixes #788 

This PR implements `Redeem Composer`, `Redeemer proxy contract` and integration of redeem flow with redeem composer. 
These contracts are similar to OST composer and Staker proxy, however there are few deviations. 

1.  `stakeRequest` mapping in the OST composer contract seems redundant. I haven't added this storage in Redeem composer contract 

https://github.com/mosaicdao/mosaic-contracts/blob/45023774d47ab8322bfb8cb495175fdf62b8a509/contracts/gateway/OSTComposer.sol#L83

2. Bounty in co-gateway contract is in base token instead of EIP20 token. `AceeptRedeemRequest`  is a payable function.

### CODE REVIEW

File: contracts/gateway/RedeemComposer.sol [hunk 1]
```diff
@@ -0,0 +1,447 @@
...
+        require(
+            utilityToken.transfer(address(redeemerProxy), _amount),
+            "Redeem amount must be transferred to the redeem proxy."
+        );
+
+        uint256 bounty = _cogateway.bounty();
+        require(
+            bounty == msg.value,
+            'Bounty amount must be received.'
+        );
```
[the remark this answers is not in the document]
Comment (sarvesh-ost) [reply contracts/gateway/RedeemComposer.sol]: Done 👍

File: contracts/gateway/RedeemPool.sol [hunk 2]
```diff
@@ -0,0 +1,447 @@
+pragma solidity ^0.5.0;
+
+import "../utilitytoken/contracts/organization/contracts/OrganizationInterface.sol";
+import "../lib/Mutex.sol";
+import "../utilitytoken/contracts/organization/contracts/Organized.sol";
+import "./EIP20CoGatewayInterface.sol";
+import "./RedeemerProxy.sol";
```
Comment (deepesh-kn) [inline contracts/gateway/RedeemPool.sol]: Please move the imports after the copyright sections

File: contracts/gateway/RedeemPool.sol [hunk 3]
```diff
@@ -0,0 +1,447 @@
...
+        address indexed redeemer,
+        bytes32 redeemRequestHash
+    );
+
+    /** Emitted whenever a reject redeem is called. */
+    event RedeemRejected(
+        address indexed redeemer,
+        bytes32 redeemRequestHash
+    );
+
```
Comment (deepesh-kn) [inline contracts/gateway/RedeemPool.sol]: Add a new line here.

File: contracts/gateway/RedeemPool.sol [hunk 4]
```diff
@@ -0,0 +1,447 @@
...
+            _gasLimit,
+            _nonce,
+            msg.sender,
+            address(redeemerProxy),
+            address(_cogateway),
+            redeemRequestHash_
+        );
+
+    }
+
```
Comment (deepesh-kn) [inline contracts/gateway/RedeemPool.sol]: Remove a new line.

File: contracts/gateway/RedeemPool.sol [hunk 5]
```diff
@@ -0,0 +1,447 @@
...
+            _gasLimit,
+            _nonce,
+            _redeemer,
+            address(_cogateway)
+        );
+
+        require(
+            redeemRequestHash == redeemRequestHashes[_redeemer][address(_cogateway)],
+            'Redeem request must exists.'
+        );
```
Comment (deepesh-kn) [inline contracts/gateway/RedeemPool.sol]: Should this `require` and at line 299 be placed in `removeRedeemRequest` function?
What do you think?

File: contracts/gateway/RedeemPool.sol [hunk 6]
```diff
@@ -0,0 +1,447 @@
...
+        redeemRequestHash_ = keccak256(
+            abi.encodePacked(
+                byte(0x19), // the initial 0x19 byte
+                byte(0x01), // the version byte for structured data
+                DOMAIN_SEPARATOR,
+                hash
+            )
+        );
+    }
+
```
Comment (deepesh-kn) [inline contracts/gateway/RedeemPool.sol]: Add a new line.

File: contracts/gateway/RedeemerProxy.sol [hunk 7]
```diff
@@ -0,0 +1,161 @@
+pragma solidity ^0.5.0;
+
+import "./EIP20CoGatewayInterface.sol";
+import "../lib/Mutex.sol";
```
Comment (deepesh-kn) [inline contracts/gateway/RedeemerProxy.sol]: Please move the import statements after copyright sections.

File: contracts/gateway/RedeemerProxy.sol [hunk 8]
```diff
@@ -0,0 +1,161 @@
...
+     *
+     * @param _owner The owner that this proxy is deployed for.
+     */
+    constructor(address payable _owner)
+        public
+    {
+        composer = msg.sender;
+        owner = _owner;
+    }
+    /**
```
Comment (deepesh-kn) [inline contracts/gateway/RedeemerProxy.sol]: Add a new line.

File: test_integration/03_redeem_and_unstake/02_redeem_and_unstake_with_redeem_composer.js [hunk 9]
```diff
@@ -0,0 +1,306 @@
...
+    await requestRedeemAssertion.verify(
+      requestEvent,
+      redeemRequest,
+      transactionFeeInProgress,
+      initialBalancesBeforeRedeem,
+    );
+  });
+
+
+  it('aceept redeem', async () => {
```
Comment (deepesh-kn) [inline test_integration/03_redeem_and_unstake/02_redeem_and_unstake_with_redeem_composer.js]:
```suggestion
  it('accept redeem', async () => {
```

File: test_integration/03_redeem_and_unstake/utils/accept_redeem_assertion.js [hunk 10]
```diff
@@ -0,0 +1,233 @@
...
+  constructor(redeemPool, cogateway, ostPrime, web3) {
+    this.redeemPool = redeemPool;
+    this.cogateway = cogateway;
+    this.token = ostPrime;
+    this.web3 = web3;
+  }
+
+  /**
+     * This verifies event and balances.
+     * @param {Object} event Event object after decoding.
```
Comment (deepesh-kn) [inline test_integration/03_redeem_and_unstake/utils/accept_redeem_assertion.js]: Please correct the indentations of all the documentations in this file

File: test_integration/03_redeem_and_unstake/utils/request_redeem_assertion.js [hunk 11]
```diff
@@ -0,0 +1,224 @@
...
+ * @property {BN} balances.baseToken.redeemer Base token(ETH) balance of redeemer.
+ */
+
+/**
+ * Class to assert event and balances after request Redeem.
+ */
+class RequestRedeemAssertion {
+  /**
+     * Constructor.
+     * @param {Object} redeemPool Truffle redeemPool instance.
```
Comment (deepesh-kn) [inline test_integration/03_redeem_and_unstake/utils/request_redeem_assertion.js]: Please correct the indentations of the documentation sections in this file.

File: contracts/gateway/EIP20CoGatewayInterface.sol [hunk 12]
```diff
@@ -0,0 +1,88 @@
...
+// http://www.simpletoken.org/
+//
+// ----------------------------------------------------------------------------
+
+import "../lib/EIP20Interface.sol";
+
+
+interface EIP20CoGatewayInterface {
+
+
```
Comment (deepesh-kn) [inline contracts/gateway/EIP20CoGatewayInterface.sol]: Remove new line.

File: contracts/gateway/EIP20CoGatewayInterface.sol [hunk 13]
```diff
@@ -0,0 +1,88 @@
...
+        uint256 _amount,
+        address _beneficiary,
+        uint256 _gasPrice,
+        uint256 _gasLimit,
+        uint256 _nonce,
+        bytes32 _hashLock
+    )
+    external
+    payable
+    returns (bytes32 messageHash_);
```
Comment (deepesh-kn) [inline contracts/gateway/EIP20CoGatewayInterface.sol]: Fix indentations.

File: contracts/gateway/RedeemerProxy.sol [hunk 14]
```diff
@@ -0,0 +1,161 @@
...
+// http://www.simpletoken.org/
+//
+// ----------------------------------------------------------------------------
+
+contract RedeemerProxy is Mutex {
+
+    /* Storage */
+
+    /** The composer that deployed this contract. */
+    address public composer;
```
Comment (deepesh-kn) [inline contracts/gateway/RedeemerProxy.sol]: Should we still call it as a composer?

Comment (sarvesh-ost) [reply contracts/gateway/RedeemerProxy.sol]: Renamed to redeemPool

**Pull request merged**